feat(slack): Support slack incoming webhooks

costimuraru opened the pull request:
Fixes https://github.com/spinnaker/spinnaker/issues/2103

This PR adds support for tokens generated for Slack Incoming web hooks [1], which follow this format: `T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX` and target a different Slack API.

This is backward compatible - still works with tokens that rely on chat.postMessage API [2]. 

The implementation is inspired from ansible's slack implementation [3].

Feedback is welcomed. We're using this internally in our environment with success.

![ss](https://user-images.githubusercontent.com/952836/33286845-51fb0360-d3bf-11e7-8faa-faff26d564aa.jpg)



<img width="642" alt="screenshot 2017-11-25 21 07 07" src="https://user-images.githubusercontent.com/952836/33233799-a1668fc4-d224-11e7-807f-d2832dabd6f9.png">

<img width="534" alt="screenshot 2017-11-25 20 49 25" src="https://user-images.githubusercontent.com/952836/33233719-14cd9e78-d223-11e7-88aa-cdc0fe6a548b.png">


[1] https://api.slack.com/incoming-webhooks
[2] https://api.slack.com/methods/chat.postMessage
[3] https://github.com/ansible/ansible-modules-extras/blob/devel/notification/slack.py#L236

anotherchrisberry commented:
I really like the look of this!

It looks like markdown formatting isn't rendering (looking at "*Instructions*" in the screenshot). Reading the [Slack documentation](https://api.slack.com/docs/message-attachments#text), it seems like markdown should be supported, so not sure what's happening there. If you can figure that out (or at least verify markdown is still rendering when no `slack.token` is in the config), I'm happy to merge this.

costimuraru commented:
Thanks for the excellent feedback @anotherchrisberry!

Indeed the markdown was not applied on the message, because there was a missing field:
```java
List<String> mrkdwn_in = ["text"]
```

I've added it now and also added tests to make sure we don't miss it again in the future. Did a test locally and it seems to work fine. I'll also deploy this build on our spinnaker infra in a couple of hours and will let you how it goes.

costimuraru commented:
@anotherchrisberry, tested this new echo version on our environment and it works quite nicely with the markdown! I've updated the screenshot in the PR description, where you can see that `Instructions` is now **bold**.
